File: specter/src/core/parser/parse.ts
/**
 * spec-parse: YAML-to-SpecAST parser.
 *
 * Pure function. No CLI deps, no I/O.
 * Validates .spec.yaml content against the canonical JSON Schema
 * and returns a typed SpecAST or structured parse errors.
 *
 * @spec spec-parse
 */

import Ajv2020, { type ErrorObject } from 'ajv/dist/2020.js';
import addFormats from 'ajv-formats';
import { parseDocument, YAMLParseError } from 'yaml';
import specSchema from '../schema/spec-schema.json' with { type: 'json' };
import type { SpecAST, SpecDocument } from '../schema/types.js';
import type { ParseError, ParseResult } from './errors.js';

// C-01: Validate against canonical JSON Schema
// C-07: Collect all errors (allErrors: true)
// C-03: Reject unknown fields (the schema has additionalProperties: false)
const ajv = new Ajv2020({ allErrors: true, strict: false });
addFormats(ajv);
const validate = ajv.compile<SpecDocument>(specSchema);

/**
 * Convert an Ajv error path to a human-readable JSON path.
 * e.g., "/spec/constraints/0/id" -> "spec.constraints[0].id"
 */
function ajvPathToJsonPath(instancePath: string): string {
  return instancePath
    .replace(/^\//, '')
    .replace(/\//g, '.')
    .replace(/\.(\d+)\./g, '[$1].')
    .replace(/\.(\d+)$/, '[$1]');
}

/**
 * Build a ParseError from an Ajv ErrorObject.
 * C-02: Include field path. Line numbers are added in a later pass via YAML source mapping.
 */
function ajvErrorToParseError(error: ErrorObject, yamlLineMap: Map<string, number>): ParseError {
  const keyword = error.keyword;
  let path: string;

  if (keyword === 'required') {
    const missingProp = error.params?.missingProperty as string | undefined;
    const parentPath = ajvPathToJsonPath(error.instancePath);
    path = parentPath ? `${parentPath}.${missingProp}` : (missingProp ?? '');
  } else if (keyword === 'additionalProperties') {
    const extra = error.params?.additionalProperty as string | undefined;
    const parentPath = ajvPathToJsonPath(error.instancePath);
    path = parentPath ? `${parentPath}.${extra}` : (extra ?? '');
  } else {
    path = ajvPathToJsonPath(error.instancePath);
  }

  const line = yamlLineMap.get(path);

  return {
    path,
    type: keyword,
    message: error.message ?? `Validation failed: ${keyword}`,
    ...(line !== undefined ? { line } : {}),
  };
}

/**
 * Build a line number map from a YAML document.
 * Maps JSON paths (e.g., "spec.id") to their YAML source line numbers.
 */
function buildLineMap(yamlContent: string): Map<string, number> {
  const lineMap = new Map<string, number>();

  try {
    const doc = parseDocument(yamlContent);
    if (!doc.contents) return lineMap;

    function walk(node: unknown, pathParts: string[]): void {
      if (!node || typeof node !== 'object') return;

      if ('items' in node && Array.isArray((node as { items: unknown[] }).items)) {
        const mapNode = node as {
          items: Array<{ key?: { value?: string; range?: number[] }; value?: unknown }>;
        };
        for (const pair of mapNode.items) {
          if (pair.key && 'value' in pair.key) {
            const key = String(pair.key.value);
            const currentPath = [...pathParts, key].join('.');
            if (pair.key.range) {
              const line = yamlContent.slice(0, pair.key.range[0]).split('\n').length;
              lineMap.set(currentPath, line);
            }
            if (pair.value && typeof pair.value === 'object') {
              if (
                'items' in pair.value &&
                Array.isArray((pair.value as { items: unknown[] }).items)
              ) {
                const seqItems = (pair.value as { items: unknown[] }).items;
                if (
                  seqItems.length > 0 &&
                  seqItems[0] &&
                  typeof seqItems[0] === 'object' &&
                  'items' in (seqItems[0] as Record<string, unknown>)
                ) {
                  // Array of maps
                  for (let i = 0; i < seqItems.length; i++) {
                    walk(seqItems[i], [...pathParts, key, String(i)]);
                  }
                }
              } else {
                walk(pair.value, [...pathParts, key]);
              }
            }
          }
        }
      }
    }

    walk(doc.contents, []);
  } catch {
    // If line mapping fails, we still return errors — just without line numbers
  }

  return lineMap;
}

/**
 * Parse a YAML string into a validated SpecAST.
 *
 * C-08: Pure function — no I/O, no CLI deps.
 * C-05: Handles YAML syntax errors gracefully.
 * C-06: YAML anchors/aliases resolved by the yaml library.
 * C-01: Validates against canonical JSON Schema.
 * C-07: Collects all validation errors.
 * C-04: Returns typed SpecAST on success.
 *
 * @param yamlContent - Raw YAML string content
 * @returns ParseResult containing either the validated SpecAST or an array of ParseErrors
 */
export function parseSpec(yamlContent: string): ParseResult<SpecAST> {
  // Step 1: Parse YAML (C-05: graceful syntax error handling, C-06: anchors resolved)
  let parsed: unknown;
  try {
    const doc = parseDocument(yamlContent, { merge: true });
    if (doc.errors.length > 0) {
      // C-02: Include line numbers from YAML errors
      const errors: ParseError[] = doc.errors.map((err: YAMLParseError) => ({
        path: '',
        type: 'yaml_syntax',
        message: err.message,
        ...(err.linePos ? { line: err.linePos[0].line, column: err.linePos[0].col } : {}),
      }));
      return { ok: false, errors };
    }
    parsed = doc.toJS();
  } catch (err) {
    // C-05: Never crash — return structured error
    if (err instanceof YAMLParseError) {
      return {
        ok: false,
        errors: [
          {
            path: '',
            type: 'yaml_syntax',
            message: err.message,
            ...(err.linePos ? { line: err.linePos[0].line, column: err.linePos[0].col } : {}),
          },
        ],
      };
    }
    return {
      ok: false,
      errors: [
        {
          path: '',
          type: 'yaml_syntax',
          message: err instanceof Error ? err.message : 'Unknown YAML parse error',
        },
      ],
    };
  }

  // Step 2: Validate against JSON Schema (C-01, C-03, C-07)
  const valid = validate(parsed);
  if (!valid && validate.errors) {
    const lineMap = buildLineMap(yamlContent);
    const errors: ParseError[] = validate.errors.map((err) => ajvErrorToParseError(err, lineMap));
    return { ok: false, errors };
  }

  // Step 3: Return typed SpecAST (C-04)
  const doc = parsed as SpecDocument;
  return { ok: true, value: doc.spec };
}

/**
 * Parse multiple YAML strings. Convenience wrapper over parseSpec.
 *
 * @param entries - Array of [filename, yamlContent] pairs
 * @returns Map of filename to ParseResult
 */
export function parseSpecs(entries: Array<[string, string]>): Map<string, ParseResult<SpecAST>> {
  const results = new Map<string, ParseResult<SpecAST>>();
  for (const [filename, content] of entries) {
    results.set(filename, parseSpec(content));
  }
  return results;
}

File: specter/internal/parser/parse.go
// Package parser implements spec-parse: YAML-to-SpecAST parser.
//
// Pure functions. No CLI deps, no I/O beyond what's passed in.
//
// @spec spec-parse
package parser

import (
	"embed"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/Hanalyx/specter/internal/schema"
	"github.com/santhosh-tekuri/jsonschema/v6"
	"gopkg.in/yaml.v3"
)

//go:embed spec-schema.json
var schemaFS embed.FS

// SchemaBytes returns the raw embedded spec-schema.json. Used by packages that
// need to walk the schema (e.g., internal/explain) without re-embedding the file.
func SchemaBytes() ([]byte, error) {
	return schemaFS.ReadFile("spec-schema.json")
}

// ParseError represents a validation error with path and optional line info.
type ParseError struct {
	Path    string `json:"path"`
	Type    string `json:"type"`
	Message string `json:"message"`
	Line    int    `json:"line,omitempty"`
	Column  int    `json:"column,omitempty"`
}

func (e ParseError) Error() string {
	if e.Line > 0 {
		return fmt.Sprintf("[%s] %s:%d: %s", e.Type, e.Path, e.Line, e.Message)
	}
	return fmt.Sprintf("[%s] %s: %s", e.Type, e.Path, e.Message)
}

// ParseResult holds the outcome of parsing a spec.
type ParseResult struct {
	OK     bool            `json:"ok"`
	Value  *schema.SpecAST `json:"value,omitempty"`
	Errors []ParseError    `json:"errors,omitempty"`
}

var compiledSchema *jsonschema.Schema

func init() {
	data, err := schemaFS.ReadFile("spec-schema.json")
	if err != nil {
		panic(fmt.Sprintf("failed to read embedded schema: %v", err))
	}

	var schemaDoc interface{}
	if err := json.Unmarshal(data, &schemaDoc); err != nil {
		panic(fmt.Sprintf("failed to parse schema JSON: %v", err))
	}

	c := jsonschema.NewCompiler()
	// Turn on format assertions so `format: date`, `format: email`, etc. are
	// validated, not just annotated. Without this, `approval_date: "not-a-date"`
	// would pass schema validation since draft 2020-12 treats format as
	// annotation-only by default.
	c.AssertFormat()
	if err := c.AddResource("spec-schema.json", schemaDoc); err != nil {
		panic(fmt.Sprintf("failed to add schema resource: %v", err))
	}

	compiled, err := c.Compile("spec-schema.json")
	if err != nil {
		panic(fmt.Sprintf("failed to compile schema: %v", err))
	}
	compiledSchema = compiled
}

// ParseSpec parses YAML content into a validated SpecAST.
//
// C-01: Validates against canonical JSON Schema.
// C-02: Reports errors with paths.
// C-03: Rejects unknown fields (additionalProperties: false in schema).
// C-04: Returns typed SpecAST on success.
// C-05: Handles YAML syntax errors gracefully.
// C-06: YAML anchors resolved by yaml.v3.
// C-07: Collects all validation errors.
// C-08: Pure function.
// maxSpecBytes caps input size before YAML parsing to prevent anchor-expansion
// DoS ("billion laughs") when Specter runs in CI on externally-sourced spec files.
const maxSpecBytes = 1 << 20 // 1 MB

func ParseSpec(yamlContent string) ParseResult {
	if len(yamlContent) > maxSpecBytes {
		return ParseResult{Errors: []ParseError{{
			Path:    "",
			Type:    "file_too_large",
			Message: fmt.Sprintf("spec file exceeds %d byte limit (%d bytes)", maxSpecBytes, len(yamlContent)),
		}}}
	}

	// Step 1: Parse YAML (C-05, C-06)
	var raw interface{}
	if err := yaml.Unmarshal([]byte(yamlContent), &raw); err != nil {
		pe := ParseError{
			Path:    "",
			Type:    "yaml_syntax",
			Message: err.Error(),
		}
		// Extract line info from yaml error
		if yamlErr, ok := err.(*yaml.TypeError); ok {
			pe.Message = yamlErr.Error()
		}
		// Try to get line number from the error string
		line := extractLineFromYAMLError(err.Error())
		if line > 0 {
			pe.Line = line
		}
		return ParseResult{OK: false, Errors: []ParseError{pe}}
	}

	// Convert YAML map keys to string (yaml.v3 produces map[string]interface{})
	normalized := normalizeYAML(raw)

	// Step 2: Validate against JSON Schema (C-01, C-03, C-07)
	err := compiledSchema.Validate(normalized)
	if err != nil {
		validationErr, ok := err.(*jsonschema.ValidationError)
		if !ok {
			return ParseResult{OK: false, Errors: []ParseError{
				{Path: "", Type: "validation", Message: err.Error()},
			}}
		}

		errors := flattenValidationErrors(validationErr)
		if len(errors) > 0 {
			return ParseResult{OK: false, Errors: errors}
		}
	}

	// Step 3: Unmarshal into typed struct (C-04)
	var doc schema.SpecDocument
	if err := yaml.Unmarshal([]byte(yamlContent), &doc); err != nil {
		return ParseResult{OK: false, Errors: []ParseError{
			{Path: "", Type: "unmarshal", Message: err.Error()},
		}}
	}

	// Step 4: Structural cross-reference validation.
	// JSON Schema can enforce that references_constraints entries MATCH the
	// ^C-\d{2,}$ pattern, but cannot assert that the referenced constraint
	// actually exists in this spec. That's a parse-level structural error —
	// same class as "id doesn't match its pattern" — so we catch it here
	// rather than waiting for specter check downstream.
	if xrefErrors := validateReferencedConstraints(&doc.Spec); len(xrefErrors) > 0 {
		return ParseResult{OK: false, Errors: xrefErrors}
	}

	return ParseResult{OK: true, Value: &doc.Spec}
}

// validateReferencedConstraints asserts every AC's references_constraints
// entries point at a constraint declared in the same spec. Returns a ParseError
// per dangling reference.
func validateReferencedConstraints(spec *schema.SpecAST) []ParseError {
	known := make(map[string]bool, len(spec.Constraints))
	for _, c := range spec.Constraints {
		known[c.ID] = true
	}

	var errs []ParseError
	for _, ac := range spec.AcceptanceCriteria {
		for _, ref := range ac.ReferencesConstraints {
			if !known[ref] {
				errs = append(errs, ParseError{
					Type:    "dangling_reference",
					Path:    fmt.Sprintf("spec.acceptance_criteria[%s].references_constraints", ac.ID),
					Message: fmt.Sprintf("references constraint %q which is not declared in this spec (declared constraints: %v)", ref, sortedKeys(known)),
				})
			}
		}
	}
	return errs
}

func sortedKeys(m map[string]bool) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	// Simple sort — small lists, no need for sort package import overhead
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0 && keys[j-1] > keys[j]; j-- {
			keys[j-1], keys[j] = keys[j], keys[j-1]
		}
	}
	return keys
}

// flattenValidationErrors converts nested jsonschema validation errors into flat ParseErrors.
func flattenValidationErrors(ve *jsonschema.ValidationError) []ParseError {
	var errors []ParseError

	if len(ve.Causes) == 0 {
		path := locationToPath(ve.InstanceLocation)
		errType := extractErrorType(ve)
		errors = append(errors, ParseError{
			Path:    path,
			Type:    errType,
			Message: humanizeError(errType, path, ve.Error()), // C-09: human-readable messages
		})
	}

	for _, cause := range ve.Causes {
		errors = append(errors, flattenValidationErrors(cause)...)
	}

	return errors
}

// locationToPath converts a jsonschema v6 InstanceLocation ([]string) to dot notation.
func locationToPath(parts []string) string {
	if len(parts) == 0 {
		return ""
	}
	var result []string
	for _, part := range parts {
		if len(part) > 0 && part[0] >= '0' && part[0] <= '9' {
			if len(result) > 0 {
				result[len(result)-1] = result[len(result)-1] + "[" + part + "]"
			}
		} else {
			result = append(result, part)
		}
	}
	return strings.Join(result, ".")
}

// extractErrorType determines the type of validation error.
func extractErrorType(ve *jsonschema.ValidationError) string {
	msg := ve.Error()
	switch {
	case strings.Contains(msg, "missing property"):
		return "required"
	case strings.Contains(msg, "additional properties"):
		return "additionalProperties"
	case strings.Contains(msg, "pattern"):
		return "pattern"
	case strings.Contains(msg, "enum"):
		return "enum"
	case strings.Contains(msg, "expected"):
		return "type"
	case strings.Contains(msg, "minimum") || strings.Contains(msg, "minItems"):
		return "minItems"
	default:
		return "validation"
	}
}

// normalizeYAML ensures all map keys are strings (required by JSON Schema validation).
func normalizeYAML(v interface{}) interface{} {
	switch val := v.(type) {
	case map[string]interface{}:
		result := make(map[string]interface{})
		for k, v := range val {
			result[k] = normalizeYAML(v)
		}
		return result
	case map[interface{}]interface{}:
		result := make(map[string]interface{})
		for k, v := range val {
			result[fmt.Sprintf("%v", k)] = normalizeYAML(v)
		}
		return result
	case []interface{}:
		result := make([]interface{}, len(val))
		for i, v := range val {
			result[i] = normalizeYAML(v)
		}
		return result
	default:
		return v
	}
}

func extractLineFromYAMLError(msg string) int {
	// yaml.v3 errors contain "line N:" patterns
	var line int
	_, _ = fmt.Sscanf(msg, "yaml: line %d:", &line)
	return line
}
